### PR TITLE
fix: redeploy api when function name changes

### PR DESCRIPTION
**Pull request by ShreyaGangishetty**
*Issue #, if available:*
#634 
*Description of changes:*
Updated the code to take function name when redeploying api
This PR resolves only `Ref` and `Fn::Sub` intrinsics for parameter values

*Description of how you validated changes:*
added a test which updates the deployment hash when functionname changes

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Verify transformed template deploys and application functions as expected

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/translator/translator.py [hunk 1]
```diff
@@ -31,6 +31,39 @@ def __init__(self, managed_policy_map, sam_parser, plugins=None):
         self.plugins = plugins
         self.sam_parser = sam_parser
 
+    function_names = dict()
+
+    def _get_function_names(self, resource_dict, parameter_values):
```
Comment (ShreyaGangishetty) [inline samtranslator/translator/translator.py]: This function appends all the function names of an API as an API can have multiple functions. If atleast one function name changes, the api will be redeployed.

File: samtranslator/translator/translator.py [hunk 2]
```diff
@@ -31,6 +31,39 @@ def __init__(self, managed_policy_map, sam_parser, plugins=None):
...
+        if resource_dict.get("Type", "").strip() == 'AWS::Serverless::Function':
+            if resource_dict.get('Properties', "").get('Events', ""):
+                events = list(resource_dict.get('Properties', "").get('Events', "").values())
+                for item in events:
+                    # If the function event type is `Api`then gets the function name and
+                    # adds to the function_names dict with key as the api_name and value as the function_name
+                    if item.get('Type', "") == 'Api' and item.get('Properties', None) and item.get('Properties', "")\
```
Comment (praneetap) [inline samtranslator/translator/translator.py]: You don't need ", None) in 'get', that's the default behavior

File: samtranslator/translator/translator.py [hunk 3]
```diff
@@ -31,6 +31,39 @@ def __init__(self, managed_policy_map, sam_parser, plugins=None):
...
+                        if api_name:
+                            function_name = resource_dict.get('Properties', "").get("FunctionName", None)
+                            if type(function_name) == str:
+                                # All the function_names for associated functions will be concatenated
+                                self.function_names[api_name] = self.function_names.get(api_name, "") +\
+                                                            resource_dict.get('Properties', "").get('FunctionName', "")
+                            # Resolving intrinsics and gets the function name
+                            # This gets the function names only for Ref and Fn::Sub intrinsics
+                            elif type(function_name) == dict:
+                                if function_name.get('Ref'):
```
Comment (ShreyaGangishetty) [inline samtranslator/translator/translator.py]: This case works for `!Ref FunctionParameter`

File: samtranslator/translator/translator.py [hunk 4]
```diff
@@ -31,6 +31,39 @@ def __init__(self, managed_policy_map, sam_parser, plugins=None):
...
+                                    self.function_names[api_name] = self.function_names.get(api_name, "") + \
+                                        parameter_values.get(function_name.get('Ref'))
+                                if function_name.get('Fn::Sub'):
```
Comment (ShreyaGangishetty) [inline samtranslator/translator/translator.py]: This resolves `Fn::Sub: FunctionName` to a string

File: samtranslator/model/apigateway.py [hunk 5]
```diff
@@ -99,7 +101,12 @@ def make_auto_deployable(self, stage, openapi_version=None, swagger=None, domain
             hash_input.append(str(openapi_version))
         if domain:
             hash_input.append(json.dumps(domain))
-
+        if redeploy_restapi_parameters:
+            function_names = redeploy_restapi_parameters.get('function_names')
+        else:
+            function_names = None
+        if function_names and function_names.get(self.logical_id[:-10], None):
```
Comment (keetonian) [inline samtranslator/model/apigateway.py]: Can you remind me again why `self.logical_id[:-10]` is needed here? Maybe add a comment about what you're trying to do?

File: samtranslator/translator/translator.py [hunk 6]
```diff
@@ -31,6 +31,36 @@ def __init__(self, managed_policy_map, sam_parser, plugins=None):
         self.plugins = plugins
         self.sam_parser = sam_parser
 
+    def _get_function_names(self, resource_dict, parameter_values):
```
Comment (keetonian) [inline samtranslator/translator/translator.py]: Is it not possible to use the intrinsic resolver? This looks like a lot of duplicated effort.

Intrinsic resolver does additional things, like resolving "!Sub ${parameter}-name", which this doesn't appear to handle.

File: samtranslator/model/apigateway.py (hunk 5 above)
Comment (ShreyaGangishetty) [reply samtranslator/model/apigateway.py]: Sure. I will add. self.logical_id has the api_name and the word "deployment". I am removing the word "deployment" and just getting the api_name to search for corresponding string of function_name/names

File: samtranslator/translator/translator.py (hunk 6 above)
Comment (ShreyaGangishetty) [reply samtranslator/translator/translator.py]: intrinsics resolver is not being called from this file. I will check and get back if i can call intrinsics resolver here.

File: samtranslator/translator/translator.py (hunk 2 above)
Comment (ShreyaGangishetty) [reply samtranslator/translator/translator.py]: Thank you. I have updated the code without ""/None

**Comment by patrickgreenwell** [comment]
This is still an issue if you reference the RestApiId via a Ref and have a FunctionName with a Sub in it. The OrderDict object seems to be more specific than needs be and templates can still fall into the else statement under the right conditions.  I've made a branch for a change that will resolve the issue.

**Pull request merged**